### DISCUSSION TITLE
Artificial horizon

### DISCUSSION THREAD

**Discussion by defilante** [dawigit/picoclock#12] [Ideas]
Hi Dawigit. 
First of all I would like to compliment you because the Picoclok is really beautiful. 
The Picoclock shows well the graphic characteristics and calculation capacity of the Rp 2040. I have been trying for some time to make a simple artificial horizon with MicroPython but I am not capable. 
When I saw the Picoclock using acceleration sensors so well I was amazed. 
Have you ever thought about making an attitude indicator? 
It is certainly much simpler than such a complex watch and would be very useful for ultralights and experimental airplanes.
Happy Holidays

**Comment by dawigit** [comment]
Hi defilante,
attitude indicator? or altitude indicator?
You'd need an external sensor for the altitude (preassure?).
I've never done any aviation stuff.

**Comment by defilante** [comment]
Hi Dawigit, I really meant an artificial horizon.
Picoclock already knows its position in space. 
I think it would be enough to change the graphics and make a half orange and half blue panel, like all the artificial horizons and start the level position from the vertical instead of horizontal.
So you can mount it on a panel.
I just can't get there.
I think Dawigit can do it in half an hour. :-)


Here an example with Arduino
https://www.youtube.com/watch?v=ta71EIkgIqo
https://forum.arduino.cc/t/artificial-horizon-display-using-a-gyro/402237/16
https://www.moreware.org/wp/blog/2023/11/17/19966/
also simple
https://youtu.be/mU1mT4dsyRU

**Reply by dawigit** [reply]
Hi Andrea.
Where did you get stuck? (It's not trivial!)
picoclock is not using the z-'axis' (which is no axis, actually).
And just by having a gyroscope/accelometer you do **not** know your position in space.
Actually there's no space. It just gives us the rotation about the x-axis and the y-axis. And
the z-value (not an axis rotation!).

I'd start by printing the values of acc[0-2] (x,y,z) and watch how they change.
Draw a attention line. (green short line)
Then get the length of your attention [ sqrtf(x*x+y*y) ].
Unify the attention so you can get the right sin/cos and the right angle (asin/acos).
Then in counter direction (to the small green line) I'd draw a cyan (fixed length line).
Cyan shows the middle. Now we also have all the angles.
I'd draw a line (mid angle+90°) and fill below/above.  (not trivial!)

It's much easier to simply draw 'wings' from the center showing your attention.
But that's not artificial horizon.

So a better (and faster) one could look like this one:
![att0](https://github.com/dawigit/picoclock/assets/26333559/96465b6e-3c35-4e31-9a16-5f84d250a27f)

But what's the use for that? What panel? Like those RC videos on youtube?

**Reply by defilante** [reply]
Fantastic, just remove the clock hands and put on two small wings!

**Comment by defilante** [comment]
So here's the full story.
I have a small flying school in Italy and we use Tecnam P 92 aircraft. It would be very useful to have an artificial horizon on board. 
![800px-Tecnam_P92_Echo_Super](https://github.com/dawigit/picoclock/assets/23702958/b2b35088-18d5-4544-a512-7ea8810b6753)

We have installed mechanical horizons but since these use gyroscopic masses (like a spinning top) that rotate at high speed and work on pins as thin as needles, after a few months of landing on grass runways, they fail due to jolts.
I thought about installing solid state artificial horizons but those certified according to general aviation regulations have very high prices starting from 5000 euros. But ultralights do not have the obligation to certify their instruments.
I have seen that drones and RC airplanes use very sensitive and reliable attitude indicators based on mpu 6050 chips and on Yotube I have seen some applications made with mpu 6050 and Arduino which are truly perfect for this purpose.
https://www.youtube.com/watch?v=WTkMdCqo9hg&t=2s
I mess around a little with Python and since Raspberry can be programmed with Micropython when I saw the RP2040 with accelerometers and color screen, I thought it was perfect for building a small artificial horizon resistant to the bumps of grass tracks.
I bought an Rp2040 and a 6050 chip and started studying Micropython. For now I can turn the screen on and off, but I learn quickly and in a couple of years I should do it. :-))
![20240103_163716](https://github.com/dawigit/picoclock/assets/23702958/d842ed71-f696-4bed-a103-1b5ff6dc0161)

Then I saw your Picoclok and was amazed by the graphics and functions. But you use C++ which I don't know at all. So I wouldn't know how to modify your script to simplify it and just show two wings moving on an orange and blue screen. In your opinion can this be achieved with Micropython?

**Comment by defilante** [comment]
Hi Dawigit and thanks for your attention.

**Comment by dawigit** [comment]
Ok. Thank you for the "full story".
So first of all you'd need a working attitude indicator (so nothing terrible happens in one of those "IMC"s).

https://www.youtube.com/watch?v=ohBZHKIIA_c
What about that one? (showing the degrees).

But: that's an extra project, **not** 'picoclock'.
Any additional code can be dangerous (crashes).
And people want to: download,install, yes! works!

And I think this is not an educational (sort of) project.
So maybe it's time (4U) to stop that slippery writing style, stop giving orders and commands on what to do next,
and if you want to use it in a commercial way, then you need to buy it! Or do it by yourself. Or… €$

Anyway: **this place, github.com** 
It is NOT a free programming service.
It is NOT a free it service. 
It is NOT a free 'build-me-that-thang' service.

So maybe your information about this webpage is wrong?
I also am not a free programming service. So maybe…
You got something wrong? :-)

**Reply by defilante** [reply]
Hi Dawigit, read the conclusion of my little story carefully and you will see that I only asked you for advice. 
I've never had a free meal in my life but I've offered quite a few to compensate, so don't exaggerate with your judgments. 
However, it's a great idea. 
Think how much you would like to do the script, maybe we can both make a deal.
